File: packages/kits/AriaMcpKit/rust/tests/fact_retire_audit_tests.rs
//! fact_retire_audit_tests.rs — Gate tests that prove `moot_retire_fact`
//! wires `changedBy` and `reason` into the `_storagekit_audit` row.
//!
//! Every test drives the SHIPPED path:
//!   Dispatcher::handle(tools/call "moot_retire_fact" ...) →
//!   surface.decode → SelectedKnowledgeJournalAuthority.admit →
//!   surface::execute → execute_knowledge_journal →
//!   CoordinatorKnowledgeJournalLower::retire_fact →
//!   coordinator.withdraw_kg_fact → _storagekit_audit row.
//!
//! `admit()` at rust/src/surface.rs:2337-2354 (NOT rust/src/v2/surface.rs)
//! populates `caller_binding` from
//! `registry.server_identity`; `retire_fact` at v2/knowledge_journal.rs:180
//! passes `&a.caller_binding` and `r.reason.as_deref()` to `withdraw_kg_fact`.
//! Both are what the tests assert.
//!
//! The fact_id is extracted from the v2 structured result:
//!   result["result"]["structuredContent"]["data"]["fact_id"]
//!
//! Mirrors Swift `FactRetireAuditTests.swift` in AriaMCPTests.

use aria_mcp::{dispatcher::Dispatcher, estate_registry::EstateRegistry, jsonrpc::JSONRPCRequest};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/// Drive tools/call through the full Dispatcher::handle path — the shipped
/// server route — and return the outer JSONRPC response.
fn call(dispatcher: &Dispatcher, tool: &str, arguments: serde_json::Value) -> serde_json::Value {
    let request = JSONRPCRequest::decode(&serde_json::json!({
        "jsonrpc": "2.0", "id": 1, "method": "tools/call",
        "params": { "name": tool, "arguments": arguments }
    })).expect("test JSONRPC request must decode");
    serde_json::to_value(dispatcher.handle(&request)).expect("response must serialize")
}

/// Extract the v2 structuredContent.data block from a tools/call result.
fn data(result: &serde_json::Value) -> &serde_json::Value {
    &result["result"]["structuredContent"]["data"]
}

fn is_success(result: &serde_json::Value) -> bool {
    result["result"]["isError"] == serde_json::json!(false)
}

// ---------------------------------------------------------------------------
// Test 1: reason forwarded
// ---------------------------------------------------------------------------

/// Retire a fact with reason="audit-reason-a". The audit row must carry
/// that exact reason. Before the fix, reason was always None.
#[test]
fn retire_fact_audit_row_carries_reason() {
    const SERVER_ID: &str = "aria-mcp-server";
    let mut registry = EstateRegistry::new_inmemory_bare();
    registry.server_identity = SERVER_ID.to_owned();
    // Clone the store Arc before the registry is moved into the dispatcher so
    // audit rows can be read after the call completes.
    let store = registry.default.store.clone();
    let dispatcher = Dispatcher::new(registry, "ARIA_MCP_Rust", "test", "test-serial", None);

    // File a fact to get a fact_id.
    let file_result = call(&dispatcher, "moot_file_fact", serde_json::json!({
        "subject":   "Galileo",
        "predicate": "discovered",
        "object":    "Jupiter's moons",
    }));
    assert!(is_success(&file_result), "moot_file_fact must succeed; got: {file_result}");
    let fact_id_str = data(&file_result)["fact_id"]
        .as_str()
        .expect("moot_file_fact must return fact_id in structuredContent.data")
        .to_owned();

    // Retire with an explicit reason.
    let retire_result = call(&dispatcher, "moot_retire_fact", serde_json::json!({
        "fact_id": fact_id_str,
        "reason":  "audit-reason-a",
    }));
    assert!(
        is_success(&retire_result),
        "moot_retire_fact must return isError:false; got: {retire_result}"
    );
    assert_eq!(
        data(&retire_result)["fact_id"].as_str().unwrap_or(""),
        fact_id_str.as_str(),
        "moot_retire_fact must echo the fact_id in structuredContent.data; got: {retire_result}"
    );

    // Inspect the audit row.
    let events = store
        .audit_events_for_row(&fact_id_str)
        .expect("audit read must not fail");
    assert_eq!(events.len(), 1, "exactly one audit event for a single retirement");
    let ev = &events[0];
    assert_eq!(ev.verb, "retract", "audit verb must be 'retract'");
    assert_eq!(
        ev.reason.as_deref(),
        Some("audit-reason-a"),
        "audit reason must match the caller-supplied value; got {:?}",
        ev.reason
    );
}

// ---------------------------------------------------------------------------
// Test 2: actor is server identity, not the old constant
// ---------------------------------------------------------------------------

/// The audit actor must equal the server_identity injected at registry
/// construction, not the old hard-coded constant "aria-v2-retire-fact".
#[test]
fn retire_fact_audit_actor_equals_server_identity() {
    const SERVER_ID: &str = "test-mootx01-host";
    let mut registry = EstateRegistry::new_inmemory_bare();
    registry.server_identity = SERVER_ID.to_owned();
    let store = registry.default.store.clone();
    let dispatcher = Dispatcher::new(registry, "ARIA_MCP_Rust", "test", "test-serial", None);

    let file_result = call(&dispatcher, "moot_file_fact", serde_json::json!({
        "subject":   "Copernicus",
        "predicate": "proposed",
        "object":    "heliocentrism",
    }));
    assert!(is_success(&file_result), "moot_file_fact must succeed; got: {file_result}");
    let fact_id_str = data(&file_result)["fact_id"]
        .as_str()
        .expect("moot_file_fact must return fact_id")
        .to_owned();

    call(&dispatcher, "moot_retire_fact", serde_json::json!({
        "fact_id": fact_id_str,
        "reason":  "audit-reason-b",
    }));

    let events = store
        .audit_events_for_row(&fact_id_str)
        .expect("audit read must not fail");
    assert_eq!(events.len(), 1);
    let ev = &events[0];
    assert_eq!(
        ev.actor, SERVER_ID,
        "audit actor must be server_identity '{}'; got '{}'",
        SERVER_ID, ev.actor
    );
    assert_ne!(
        ev.actor, "aria-v2-retire-fact",
        "old hard-coded constant must NOT appear as audit actor"
    );
}

// ---------------------------------------------------------------------------
// Test 3: nil reason when no reason is supplied
// ---------------------------------------------------------------------------

/// When the caller omits reason, the audit row reason must be None.
#[test]
fn retire_fact_audit_reason_is_none_when_omitted() {
    const SERVER_ID: &str = "aria-mcp-server";
    let mut registry = EstateRegistry::new_inmemory_bare();
    registry.server_identity = SERVER_ID.to_owned();
    let store = registry.default.store.clone();
    let dispatcher = Dispatcher::new(registry, "ARIA_MCP_Rust", "test", "test-serial", None);

    let file_result = call(&dispatcher, "moot_file_fact", serde_json::json!({
        "subject":   "Newton",
        "predicate": "formulated",
        "object":    "gravity",
    }));
    assert!(is_success(&file_result), "moot_file_fact must succeed; got: {file_result}");
    let fact_id_str = data(&file_result)["fact_id"]
        .as_str()
        .expect("moot_file_fact must return fact_id")
        .to_owned();

    // Retire without a reason.
    call(&dispatcher, "moot_retire_fact", serde_json::json!({
        "fact_id": fact_id_str,
    }));

    let events = store
        .audit_events_for_row(&fact_id_str)
        .expect("audit read must not fail");
    assert_eq!(events.len(), 1);
    let ev = &events[0];
    assert_eq!(ev.verb, "retract");
    assert_eq!(
        ev.reason, None,
        "audit reason must be None when no reason is supplied; got {:?}",
        ev.reason
    );
}

File: packages/kits/AriaMcpKit/Tests/AriaMCPTests/FactRetireAuditTests.swift
// FactRetireAuditTests.swift
//
// Gate tests that prove `moot_retire_fact` wires `changedBy` and `reason`
// into the `_storagekit_audit` row.
//
// Every test drives the FULL dispatch chain:
//   ToolDispatcher.dispatch(name: "moot_retire_fact", ...) → retireFact →
//   kit.retireKGFact → DrawerStore.withdrawKGFact → _storagekit_audit row.
//
// Three assertions on the single audit event that must exist:
//   1. verb == "retract"       — correct verb from withdrawKGFact
//   2. actor == server identity — changedBy is now context.serverIdentity
//   3. reason == caller reason  — reason is now forwarded from the request
//
// Each assertion stands alone, so a failure names the field that moved:
// the actor case also asserts the value is NOT the old hardcoded literal
// "aria-v2-retire-fact", and a third test covers an omitted reason.

import Testing
import Foundation
import GeniusLocusKit
import LocusKit
import PersistenceKit
import PersistenceKitInMemory
import SubstrateTypes
@testable import AriaMCP

// ---------------------------------------------------------------------------
// MARK: - Shared helpers
// ---------------------------------------------------------------------------

/// Extract the `structuredContent.data` dictionary from a dispatch result.
private func data(_ result: JSONValue) -> [String: JSONValue]? {
    result.objectValue?["structuredContent"]?.objectValue?["data"]?.objectValue
}

/// Open a bare in-memory estate for audit tests.
/// Returns the dispatcher, kit, and handle so the test can query audit rows.
private func openBareEstate(identity: String = "aria-mcp-server")
    async throws -> (ToolDispatcher, GeniusLocusKit, EstateHandle)
{
    let kit = GeniusLocusKit()
    let owner = OwnerCredentials(ownerIdentifier: "fact-retire-audit-tests")
    let storage = InMemoryStorage(
        configuration: EstateConfiguration(estateID: UUID(), backend: .inMemory))
    _ = try await LocusKit.Estate.create(storage: storage, owner: owner)
    let handle = try await kit.open(
        storage: storage,
        owner: owner,
        identityKeyStore: InMemoryEstateIdentityKeyStore())
    let dispatcher = ToolDispatcher(kit: kit, handle: handle, serverIdentity: identity)
    return (dispatcher, kit, handle)
}

// ---------------------------------------------------------------------------
// MARK: - Test suite
// ---------------------------------------------------------------------------

@Suite("moot_retire_fact audit wiring — changedBy and reason forwarded to _storagekit_audit")
struct FactRetireAuditTests {

    // -----------------------------------------------------------------------
    // Test 1: reason forwarded
    // -----------------------------------------------------------------------

    /// Retire a fact with reason="audit-reason-a". The audit row must carry
    /// that exact reason string. Before the fix, reason was always nil.
    @Test("moot_retire_fact: audit row carries the caller-supplied reason")
    func retireFactAuditRowCarriesReason() async throws {
        let (dispatcher, kit, handle) = try await openBareEstate()

        // File a fact first.
        let fileResult = try await dispatcher.dispatch(
            name: "moot_file_fact",
            arguments: .object([
                "subject":   .string("Galileo"),
                "predicate": .string("discovered"),
                "object":    .string("Jupiter's moons"),
            ]))
        let factIDStr = try #require(data(fileResult)?["fact_id"]?.stringValue)

        // Retire with an explicit reason.
        let retireResult = try await dispatcher.dispatch(
            name: "moot_retire_fact",
            arguments: .object([
                "fact_id": .string(factIDStr),
                "reason":  .string("audit-reason-a"),
            ]))
        let retiredData = try #require(data(retireResult))
        #expect(retiredData["fact_id"]?.stringValue == factIDStr,
                "moot_retire_fact must echo the fact_id in structuredContent.data")

        // Read the audit trail and check the single row.
        let events = try await kit.auditTrail(in: handle, rowID: factIDStr)
        #expect(events.count == 1, "exactly one audit event for a single retirement")
        let ev = try #require(events.first)
        #expect(ev.verb == "retract",    "audit verb must be 'retract'")
        #expect(ev.reason == "audit-reason-a",
                "audit reason must match the caller-supplied value; got \(ev.reason as Any)")
    }

    // -----------------------------------------------------------------------
    // Test 2: actor is server identity, not the old constant
    // -----------------------------------------------------------------------

    /// The audit actor must equal the serverIdentity injected at dispatcher
    /// construction, not the old hard-coded constant "aria-v2-retire-fact".
    @Test("moot_retire_fact: audit actor equals serverIdentity, not 'aria-v2-retire-fact'")
    func retireFactAuditActorEqualsServerIdentity() async throws {
        let serverID = "test-mootx01-host"
        let (dispatcher, kit, handle) = try await openBareEstate(identity: serverID)

        let fileResult = try await dispatcher.dispatch(
            name: "moot_file_fact",
            arguments: .object([
                "subject":   .string("Copernicus"),
                "predicate": .string("proposed"),
                "object":    .string("heliocentrism"),
            ]))
        let factIDStr = try #require(data(fileResult)?["fact_id"]?.stringValue)

        _ = try await dispatcher.dispatch(
            name: "moot_retire_fact",
            arguments: .object([
                "fact_id": .string(factIDStr),
                "reason":  .string("audit-reason-b"),
            ]))

        let events = try await kit.auditTrail(in: handle, rowID: factIDStr)
        #expect(events.count == 1)
        let ev = try #require(events.first)
        #expect(ev.actor == serverID,
                "audit actor must be serverIdentity '\(serverID)'; got '\(ev.actor)'")
        #expect(ev.actor != "aria-v2-retire-fact",
                "old hard-coded constant must NOT appear as audit actor")
    }

    // -----------------------------------------------------------------------
    // Test 3: nil reason when no reason is supplied
    // -----------------------------------------------------------------------

    /// When the caller omits reason, the audit row reason must be nil (not a
    /// non-nil value from a forgotten forwarding bug).
    @Test("moot_retire_fact: audit reason is nil when no reason supplied")
    func retireFactAuditReasonIsNilWhenOmitted() async throws {
        let (dispatcher, kit, handle) = try await openBareEstate()

        let fileResult = try await dispatcher.dispatch(
            name: "moot_file_fact",
            arguments: .object([
                "subject":   .string("Newton"),
                "predicate": .string("formulated"),
                "object":    .string("gravity"),
            ]))
        let factIDStr = try #require(data(fileResult)?["fact_id"]?.stringValue)

        // Retire without a reason.
        _ = try await dispatcher.dispatch(
            name: "moot_retire_fact",
            arguments: .object([
                "fact_id": .string(factIDStr),
            ]))

        let events = try await kit.auditTrail(in: handle, rowID: factIDStr)
        #expect(events.count == 1)
        let ev = try #require(events.first)
        #expect(ev.verb == "retract")
        #expect(ev.reason == nil,
                "audit reason must be nil when no reason is supplied; got \(ev.reason as Any)")
    }
}
